### PR TITLE
Port MPCD integrator, streaming methods, and collision methods

### PR DESCRIPTION
**Pull request by mphoward**
## Description

This PR ports core MPCD integration capabilities to v4. Specifically, it reimplements the `Integrator` that extends `md.Integrator` to include a `StreamingMethod` and `CollisionMethod`. All of the existing implementations of these two classes have also been reimplemented.

~~I targeted this PR onto `trunk-minor` rather than a staging branch to try to make it easier to keep things synchronized. If we do this, I propose that we do not document the inclusion of these MPCD features in the public documentation or changelog yet. I am concerned that some aspects of the API may need to change as I add back in some of the other capabilities, and I don't want to worry about making breaking changes. I think it would be good to get all necessary features added back, do some private beta testing with it, and then complete & link in documentation.~~

Per below, this will go onto the integration branch `mpcd-v4`, so this PR still needs documentation to be linked in and built, and a changelog entry. Will take care of this after initial review is done.

## Motivation and context

This is the next big change needed to make MPCD available again in HOOMD.

## How has this been tested?

The previous unit tests have been reimplemented as best I could with `pytest`. It would be great to get some feedback on whether I've followed the level of coverage and patterns that you are aiming for before I work on additional classes & tests in subsequent PRs.

## Change log

~~I propose not to add a change at this point. I can write a set of release notes when everything is ready.~~
Will add this before merging.

## Checklist:

- [X] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [X] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [X] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by mphoward** [comment]
@joaander @b-butler if you have time, I would appreciate it if one of you could also have a look at whether I've used developer features of the new Python interface as intended before I repeat my mistakes in other classes. =) I think I figured it out in the end (& got a lot less code as a result!), but it was difficult to follow the documentation & code for it in a couple places.

**Comment by joaander** [comment]
Thanks! I'll review this soon.

Unless I am missing something, the only additional maintenance burden would involve periodically merging `trunk-minor` into the integration branch. With an integration branch, you could enable and test the `mpcd` package import and sphinx documentation at every step along the way. 

I would not want to show documentation for an API in a 4.x release that is going to drastically change in 4.x+n. I am also reluctant to import `mpcd` in any release until it is documented.

**Comment by mphoward** [comment]
> Unless I am missing something, the only additional maintenance burden would involve periodically merging `trunk-minor` into the integration branch.

Nope, that should be it. We had talked about potentially avoiding that to avoid long-lived branches parallel to the trunk ones, but I am fine doing it this way. I'll retarget this PR onto `mpcd-v4`, I had created that branch for this purpose!

**Comment by joaander** [comment]
> Nope, that should be it. We had talked about potentially avoiding that to avoid long-lived branches parallel to the trunk ones, but I am fine doing it this way. I'll retarget this PR onto `mpcd-v4`, I had created that branch for this purpose!

Thanks. Hopefully the branch will not be that long lived :)

File: hoomd/md/integrate.py
Comment (mphoward) [inline hoomd/md/integrate.py]: Just wanted to add a note because this was a spot I wanted someone to have a look at.

Without this change, I think the `_cpp_obj` created in `mpcd.Integrator` would get overwritten by this one?

My alternative was to leave this alone & call the grandparent method from `mpcd.Integrator`. This would work currently because this method doesn't do anything other than initialize `_cpp_obj` then call the parent method, but I was worried that I would need to duplicate anything that got added here in future.

Comment (b-butler) [reply hoomd/md/integrate.py]: It depends on how you called it for instance,
```python
super()._attach_hook()
# mpcd logic
```
would overwrite the MD integrator.

However, I don't feel too strongly about this, but I probably would call the grandparent method.

File: hoomd/md/integrate.py
Comment (b-butler) [inline hoomd/md/integrate.py]: Is there a reason for all this unrelated changes?

File: hoomd/mpcd/CollisionMethod.cc
Comment (b-butler) [inline hoomd/mpcd/CollisionMethod.cc]: I would call this something more descriptive than `every` `period` seems fine.

File: hoomd/mpcd/CollisionMethod.cc
Comment (b-butler) [inline hoomd/mpcd/CollisionMethod.cc]: Perhaps, `embeded_group` would be better?

File: hoomd/mpcd/StreamingMethod.cc
Comment (b-butler) [inline hoomd/mpcd/StreamingMethod.cc]: Similar comment to above.

File: hoomd/mpcd/Integrator.cc
Comment (b-butler) [inline hoomd/mpcd/Integrator.cc]: Why not add `!m_stream` to the first conditional?

File: hoomd/mpcd/pytest/test_integrator.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_integrator.py:1]: I would add tests for setting and getting stream and collision methods,  detaching, and pickling.

File: hoomd/mpcd/collide.py
Comment (b-butler) [inline hoomd/mpcd/collide.py]: I think period and `filter` make the most sense. Here, `filter` at least is the common term used when we have a feature like this.

File: hoomd/mpcd/collide.py
Comment (b-butler) [inline hoomd/mpcd/collide.py]:
```suggestion
    still conserves linear momentum, but kinetic energy is no longer
```
Try to avoid language like obviously, clearly, of course, etc.

File: hoomd/mpcd/pytest/test_collide.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_collide.py]: This is testing the integrator more than the collision method outside line 50, and should be in the integrator tests.

File: hoomd/mpcd/pytest/test_collide.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_collide.py]: Generally if we are going to test a parameter like this, we would test before and after run is called.

File: hoomd/mpcd/pytest/test_collide.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_collide.py:82]: This entire test has no asserts. There is something to a test just to check an exceptions isn't raised, but should we check that the values are set correctly?

File: hoomd/mpcd/pytest/test_collide.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_collide.py]: Once again this is testing the integrator more than the collision method.

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]: We typically test before and after calling run.

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]: Couldn't we test changing `every` here.

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]: It seems the second particle is being reflected not passing through?

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]: Same as previous question.

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]: While, this may be true (didn't check), we only run for 1 step, so how is this relevant?

File: hoomd/mpcd/pytest/test_stream.py
Comment (b-butler) [inline hoomd/mpcd/pytest/test_stream.py]:
```suggestion
            snap.mpcd.position[0] = [[4.95, -4.95, 3.85]]
```
Only one particle anyways.

**Review by b-butler — CHANGES_REQUESTED** [review]
Tried to go through all the code, and leave comments.

File: hoomd/md/integrate.py
Comment (mphoward) [reply hoomd/md/integrate.py]: No, a VSCode format-on-save did this. I can revert the changes if you want.

File: hoomd/mpcd/CollisionMethod.cc
Comment (mphoward) [reply hoomd/mpcd/CollisionMethod.cc]: I switched to `every` because `period` wasn't used elsewhere anymore, but this is not a general `trigger` either. Typically when people write up these methods they say something like "a collision was performed every 5 steps" so I thought that fit. I can switch back to `period` if you prefer though!

File: hoomd/mpcd/CollisionMethod.cc
Comment (mphoward) [reply hoomd/mpcd/CollisionMethod.cc]: Does HOOMD use the concept of a "group" elsewhere now that there are "filters" though? Maybe `embedded_particles`?

File: hoomd/md/integrate.py
Comment (joaander) [reply hoomd/md/integrate.py]: Please do. Numerous unrelated changes caused by slightly differently configured auto-formatters makes code review challenging. yapf doesn't currently allow us to auto-format a specific quote style.

If we switch to ruff in the future, then we can enforce string quote styles automatically via pre-commit.

Comment (mphoward) [reply hoomd/md/integrate.py]: Gotcha, this was probably done by black being enabled at a user level, and I think I thought pre-commit would catch it and put things back. I'll revert.

File: hoomd/mpcd/CollisionMethod.cc
Comment (joaander) [reply hoomd/mpcd/CollisionMethod.cc]: I would also prefer `period`. I find class attributes that are not nouns awkward in general.

File: hoomd/mpcd/Integrator.cc
Comment (mphoward) [reply hoomd/mpcd/Integrator.cc]: You are right that it should only be a warning if there are no integration methods and no streaming method.

In MPCD, it is fine to run with only a streaming method (pure solvent), but a warning will be issued by the parent integrator when it update gets called below because there are no integration methods. So, we enter this branch if there are no integration methods and pretend to issue the warning (`m_gave_warning = true`) to avoid that case in the parent.

This would be better solved by making the decision to issue a warning a virtual method. I wasn't sure if that was overkill though.

File: hoomd/mpcd/pytest/test_integrator.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_integrator.py:1]: Is there a file with good examples I can model these after?

File: hoomd/mpcd/CollisionMethod.cc
Comment (joaander) [reply hoomd/mpcd/CollisionMethod.cc]: Groups are still used internally. The user-facing docs allude to this in the description: https://hoomd-blue.readthedocs.io/en/v4.3.0/module-hoomd-filter.html . Namely, the filter is the object that selects particles and the group is the (not user facing, not accessible) object that represents the particles selected the last time the filter was evaluated. This is also discussed in the `FilterUpdater` docs: https://hoomd-blue.readthedocs.io/en/v4.3.0/module-hoomd-update.html#hoomd.update.FilterUpdater

File: hoomd/mpcd/collide.py
Comment (mphoward) [reply hoomd/mpcd/collide.py]: I considered `filter` but decided against it because I thought it sounded like the method would only be applied to the particles in `filter`, and not also the MPCD particles. How about `embedded_particles` in keeping with Josh's preference for nouns?

File: hoomd/mpcd/collide.py
Comment (mphoward) [reply hoomd/mpcd/collide.py]: That is a copy–paste from the original documentation. All of docstrings need review and revision, and will keep an eye out for this.

File: hoomd/mpcd/pytest/test_collide.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_collide.py]: I think I wanted to test the attach method her and this can only be done with an integrator, but lines 52–58 are not needed for this.

File: hoomd/mpcd/pytest/test_collide.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_collide.py:82]: Yes, this is a good idea.

File: hoomd/mpcd/pytest/test_collide.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_collide.py]: The intent of this test in the old code was to trigger all of the internals of the collision method to make sure they worked without error. We can consolidate this with the streaming method in a run test of the integrator though.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py]: Yes, we can. I think this might have been cumbersome with the old API.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py]: Yes good catch, the is a copy–paste error that was probably in the old code too.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py]: `every` is 3, so the particles advance 3 steps when the first run step triggers (the MPCD particles do not stay in sync with the timestep counter). I will update the comment below to make this clear.

File: hoomd/md/integrate.py
Comment (joaander) [reply hoomd/md/integrate.py]: I would also call the grandparent method. I don't anticipate adding more logic to this integrator attach hook.

File: hoomd/mpcd/Integrator.cc
Comment (joaander) [reply hoomd/mpcd/Integrator.cc]: I have been tempted to remove this warning altogether from `IntegratorTwoStep`. If a user choose to add no methods, that is their choice and hoomd's behavior is perfectly well defined in this case. Many users often use this for testing the value of forces/energies.

Comment (mphoward) [reply hoomd/mpcd/Integrator.cc]: I'm fine with that, and I would then remove this warning here too.

**Review by joaander — COMMENTED** [review]
Thanks for working on this. I don't have any detailed comments outside what @b-butler has raised.

I am confused how the unit tests are working. The
```
# if version.mpcd_built:
#     from hoomd import mpcd
```
is commented out in `hoomd/__init__.py` - yet the tests somehow successfully use `hoomd.mpcd`. How does this work?

Also, to enable pre-commit formatting and style checks, remove the `mpcd` exclusion lines from `setup.cfg` and `.pre-commit-config.yaml`. Depending on your progress you may or may not want to take these steps in this pull request or defer to the next.

Similarly, you will want to test any code snippets in the documentation (again, probably in a future PR that revises the docstrings). See https://github.com/glotzerlab/hoomd-blue/pulls?q=is%3Apr+is%3Aclosed+sybil for examples of pull requests that use sybil to test `.. code-block:: python` blocks.

**Comment by mphoward** [comment]
> I am confused how the unit tests are working. The
> 
> ```
> # if version.mpcd_built:
> #     from hoomd import mpcd
> ```
> 
> is commented out in `hoomd/__init__.py` - yet the tests somehow successfully use `hoomd.mpcd`. How does this work?

That is a great question! I didn't realize this import was commented out since the tests ran without error...

> Also, to enable pre-commit formatting and style checks, remove the `mpcd` exclusion lines from `setup.cfg` and `.pre-commit-config.yaml`. Depending on your progress you may or may not want to take these steps in this pull request or defer to the next.

I will revert the format change we noticed above, then enable this in the next PR.

> Similarly, you will want to test any code snippets in the documentation (again, probably in a future PR that revises the docstrings). See https://github.com/glotzerlab/hoomd-blue/pulls?q=is%3Apr+is%3Aclosed+sybil for examples of pull requests that use sybil to test `.. code-block:: python` blocks.

I will defer this to a later PR that improves the docstrings.

File: hoomd/mpcd/pytest/test_collide.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_collide.py]: I think we should leave this test in place since it checks that a run with only a collision method can execute a real step. This only flags if something catastrophic were to happen, and more detailed validation of the physics is currently done in C++.

The streaming methods have their own tests that run, and particle level details are checked.

I think it's still valuable to do these checks separately from each other, and that this check is primarily of the collision method (and not integrator). We should add some integration / validation type checks later that do run a full simulation with both attached to an integrator, and I'll make a note to do so.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py]: I was mistaken about this. `period` is currently readonly.

File: hoomd/mpcd/pytest/test_stream.py
Comment (mphoward) [reply hoomd/mpcd/pytest/test_stream.py]: I changed the period to 1, and increased the timestep to 0.3 so this was less confusing.

File: hoomd/mpcd/__init__.py
Comment (b-butler) [inline hoomd/mpcd/__init__.py]: Is is possible/practical to enumerate these restrictions in the documentation (sorry if you do this elsewhere and I missed it).

**Review by b-butler — APPROVED** [review]
I think the code is in a good place. The documentation will need a thorough going through before the final merge, but that can take place later when the code examples and such are in place.

File: hoomd/mpcd/__init__.py
Comment (mphoward) [reply hoomd/mpcd/__init__.py]: The main restriction is that the embedded solute particles cannot be in a rigid body. (We are going to remove this restriction as part of a funded project on this code.) I think this was explicit in the old version of the docs but got dropped out in revision. I'll add it back in `mpcd.collide.CollisionMethod`, and see if this statement can also be tightened up since it sounds more restrictive than it actually is!

**Comment by mphoward** [comment]
> The documentation will need a thorough going through before the final merge, but that can take place later when the code examples and such are in place.

Thanks for the feedback and reviews! Yes, I plan to go through all of the documentation and add examples as the final PR to this staging branch.

We also plan to add some how-tos / examples since those are sorely missing, but that can take place after the staging branch is merged in.

I will implement the doc fix above and also merge up `trunk-minor`, then this should be ready.

File: hoomd/mpcd/__init__.py
Comment (joaander) [reply hoomd/mpcd/__init__.py]: This block (and a few others) are not intended the correct number of spaces. Sphinx is rendering them as an item in a definition list: https://hoomd-blue--1655.org.readthedocs.build/en/1655/package-mpcd.html

See https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#bullet-lists.

**Comment by joaander** [comment]
Let me know when this is ready to merge.

**Comment by mphoward** [comment]
This should be ready once all the tests pass!

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**
